hashmap/robin.go

```diff
@@ -0,0 +1,252 @@
+package hashmap
+
+import (
+	g "github.com/zyedidia/generic"
+)
+
+const (
+	emptyBucket  = -1
+	resizeFactor = 2
+)
+
+type bucket[K comparable, V any] struct {
+	// psl is the probe sequence length (PSL), which is the distance value from
+	// the optimum insertion. -1 or `emptyBucket` signals a free slot.
+	// inspired from:
+	//  - https://programming.guide/robin-hood-hashing.html
+	//  - https://cs.uwaterloo.ca/research/tr/1986/CS-86-14.pdf
+	psl   int8
+	key   K
+	value V
+}
+
+// RobinMap is a hashmap that uses linear probing in combination with
+// robin hood hashing as collision strategy.
+type RobinMap[K comparable, V any] struct {
+	buckets []bucket[K, V]
+	hasher  g.HashFn[K]
+	// length stores the current inserted elements
+	length uintptr
+	// indexMask is used for a bitwise AND on the hash value,
+	// because the size of the underlying array is a power of two value
+	indexMask uintptr
+	// log2Index is the number of extra reserved bytes at the end of the array,
+	// to sparse the length check while probing.
+	// Furthermore this value is the maximum possible PSL over the hashmap,
+	// because a grow is forced if this value will raised during the insert operation.
+	log2Index int8
+}
+
+// go:inline
+func newBucketArray[K comparable, V any](capacity uintptr) []bucket[K, V] {
+	buckets := make([]bucket[K, V], capacity)
+	for i := range buckets {
+		buckets[i].psl = emptyBucket
+	}
+	return buckets
+}
+
+// NewRobinMap constructs a new map with default settings.
+func NewRobinMap[K comparable, V any]() *RobinMap[K, V] {
```

Comment:
Is it important that the key be comparable? It would be nice to be consistent between the normal map and the robin map by using an equals and hash callback instead of requiring the key to be comparable. Hopefully that will also allow deletion of the GetHasher function which uses unsafe.

Reply:
The reason why I made the key comparable is because than I can use the flat compare operator `==`
here https://github.com/zyedidia/generic/pull/41/files#diff-3e6a5b1df35ab9e622b2b5800349c79c52dfd9156768960701f5917f1038402eR85

The problem is that the golang compiler does not optimize out static callback functions. So a flat compare is significant faster than a function call.